content/livesync/models.textile

```diff
@@ -0,0 +1,357 @@
+
+The SDK supports a short time buffering for change events, facilitating short-term reordering and de-duplication. By default, this feature is not enabled. To activate it, set the @eventBufferOptions.bufferMS@ in the "@ClientOption@":#client-options to a non-zero value.
+
+When using the event buffer, change events are de-duplicated and ordered according to their "@message.id@":/api/realtime-sdk/messages#id, which corresponds to the change event's "@sequenceId@":#sequence.
+
+By default, the events in the buffer are ordered numerically if the @message.id@ can be coerced to a number. Otherwise, events will be ordered lexicographically by their @message.id@. You can specify a custom ordering based on any part of the message via the @eventBufferOptions.eventOrderer@ "client option":#client-options.
+
+h2(#optimistic-updates). Add Optimistic Updates
+
+The Models SDK supports optimistic updates, a feature that enables you to immediately render changes to your data model before the backend confirms the changes, making updates appear instantaneous.
+
+Optimistic events are used to make local, provisional changes to your data, anticipating that your backend will eventually confirm or reject these changes. The optimistic event will be processed by your "merge function":#merge and included in the local model state optimistically.
+
+Use the @model.optimistic()@ method to apply an optimistic update to your model.
+
+1. Call @model.optimistic()@ on your model with the optimistic event.
+2. Apply the corresponding change to your backend.
+3. Await the confirmation of the optimistic update from the backend, or optionally cancel the optimistic update.
+
+The @model.optimistic()@ function returns a promise resolving to two values:
+
+* A confirmation promise that resolves when the backend confirms or rejects the optimistic update.
+* A function to explicitly cancel the optimistic update.
+
+The following demonstrates an optimistic update to implementing changes in the model:
+
+```[javascript]
+// optimistically apply the changes to the model
+const [confirmation, cancel] = await model.optimistic({
+    mutationId: 'my-mutation-id',
+    name: 'addComment',
+    data: 'New comment!',
+});
+
+try {
+    // apply the changes in your backend
+    await updatePost('my-mutation-id', 'New comment!');
+    // wait for the optimistic event to be confirmed
+    await confirmation;
+} catch (err) {
+    // something went wrong, cancel the optimistic update
+    cancel();
+}
+```
+
+<aside data-type='important'>
+<p>Be aware that updating the model state in your database happens entirely through your backend endpoint.</p>
+</aside>
+
+h3(#mutation-id). Mutation ID
+
+The SDK requires a mechanism to identify when a change event received from the backend matches an optimistic event that has already been applied locally.
```

Comment:
> The SDK needs to identify when a change event received from the backend matches an optimistic event that has already been applied locally. To match confirmed and optimistic events the SDK uses a 'mutationId'. This ID can be any string, though it's commonly a "UUID":https://www.npmjs.com/package/uuid . The 'mutationId' is generated in your frontend code and assigned to the optimistic event when you call 'model.optimistic'. You then need to send this 'mutationId' to your backend as part of your API calls to mutate the model. Finally, the mutationId should be included in the confirmed event that's written to the outbox table. 

> If your backend decides to reject the the mutation, you should write the mutationId to the outbox with the 'rejected' column set to 'true'. This will cause the optimistic update to automatically rollback in the frontend client that triggered it. This rollback will be reflected in a 'rejected' `confirmation` promise returned from the call to `model.optimistic`.

> Optimistic mutations that are not confirmed or rejected within the timeout `optimisticEventOptions.timeout` defined in `ClientOptions` (default 5 seconds) will be automatically rolled back in the Model state.